File: android/src/main/kotlin/com/miracl/trust/flutter_miracl_sdk/Extensions.kt
package com.miracl.trust.flutter_miracl_sdk

import com.miracl.trust.authentication.AuthenticationException
import com.miracl.trust.configuration.ConfigurationException
import com.miracl.trust.registration.ActivationTokenException
import com.miracl.trust.registration.VerificationException
import com.miracl.trust.registration.QuickCodeException
import com.miracl.trust.registration.RegistrationException
import com.miracl.trust.session.AuthenticationSessionException
import com.miracl.trust.session.SigningSessionException
import com.miracl.trust.signing.SigningException

internal val ConfigurationException.flutterExceptionCodeRepresentation: MConfigurationExceptionCode
    get() = when (this) {
        ConfigurationException.EmptyProjectId -> MConfigurationExceptionCode.EMPTY_PROJECT_ID
        ConfigurationException.InvalidProjectUrl -> MConfigurationExceptionCode.INVALID_PROJECT_URL
    }

internal val VerificationException.flutterExceptionCodeRepresentation: MEmailVerificationExceptionCode
    get() = when (this) {
        VerificationException.EmptyUserId -> MEmailVerificationExceptionCode.EMPTY_USER_ID
        VerificationException.InvalidSessionDetails -> MEmailVerificationExceptionCode.INVALID_SESSION_DETAILS
        is VerificationException.RequestBackoff -> MEmailVerificationExceptionCode.REQUEST_BACKOFF
        is VerificationException.VerificationFail -> MEmailVerificationExceptionCode.VERIFICAITON_FAIL
    }

internal val ActivationTokenException.flutterExceptionCodeRepresentation: MActivationTokenExceptionCode
    get() = when (this) {
        ActivationTokenException.EmptyUserId -> MActivationTokenExceptionCode.EMPTY_USER_ID
        ActivationTokenException.EmptyVerificationCode -> MActivationTokenExceptionCode.EMPTY_VERIFICATION_CODE
        is ActivationTokenException.GetActivationTokenFail -> MActivationTokenExceptionCode.GET_ACTIVATION_TOKEN_FAIL
        is ActivationTokenException.UnsuccessfulVerification -> MActivationTokenExceptionCode.UNSUCCESSFUL_VERIFICATION
    }

internal val RegistrationException.flutterExceptionCodeRepresentation: MRegistrationExceptionCode
    get() = when (this) {
        RegistrationException.EmptyActivationToken -> MRegistrationExceptionCode.EMPTY_ACTIVATION_TOKEN
        RegistrationException.EmptyUserId -> MRegistrationExceptionCode.EMPTY_USER_ID
        RegistrationException.InvalidActivationToken -> MRegistrationExceptionCode.INVALID_ACTIVATION_TOKEN
        RegistrationException.InvalidPin -> MRegistrationExceptionCode.INVALID_PIN
        RegistrationException.PinCancelled -> MRegistrationExceptionCode.PIN_CANCELLED
        RegistrationException.ProjectMismatch -> MRegistrationExceptionCode.PROJECT_MISMATCH
        is RegistrationException.RegistrationFail -> MRegistrationExceptionCode.REGISTRATION_FAIL
        RegistrationException.UnsupportedEllipticCurve -> MRegistrationExceptionCode.UNSUPPORTED_ELLIPTIC_CURVE
    }

internal val AuthenticationException.flutterExceptionCodeRepresentation: MAuthenticationExceptionCode
    get() = when (this) {
        is AuthenticationException.AuthenticationFail -> MAuthenticationExceptionCode.AUTHENTICATION_FAIL
        AuthenticationException.InvalidAppLink -> MAuthenticationExceptionCode.INVALID_LINK
        AuthenticationException.InvalidAuthenticationSession -> MAuthenticationExceptionCode.INVALID_AUTHENTICATION_SESSION
        AuthenticationException.InvalidPin -> MAuthenticationExceptionCode.INVALID_PIN
        AuthenticationException.InvalidPushNotificationPayload -> MAuthenticationExceptionCode.INVALID_PUSH_NOTIFICATION_PAYLOAD
        AuthenticationException.InvalidQRCode -> MAuthenticationExceptionCode.INVALID_QRCODE
        AuthenticationException.InvalidUserData -> MAuthenticationExceptionCode.INVALID_USER_DATA
        AuthenticationException.PinCancelled -> MAuthenticationExceptionCode.PIN_CANCELLED
        AuthenticationException.Revoked -> MAuthenticationExceptionCode.REVOKED
        AuthenticationException.UnsuccessfulAuthentication -> MAuthenticationExceptionCode.UNSUCCESSFUL_AUTHENTICATION
        AuthenticationException.UserNotFound -> MAuthenticationExceptionCode.USER_NOT_FOUND
        AuthenticationException.InvalidCrossDeviceSession -> MAuthenticationExceptionCode.INVALID_CROSS_DEVICE_SESSION
    }

internal val QuickCodeException.flutterExceptionCodeRepresentation: MQuickCodeExceptionCode
    get() = when (this) {
        is QuickCodeException.GenerationFail -> MQuickCodeExceptionCode.GENERATION_FAIL
        QuickCodeException.InvalidPin -> MQuickCodeExceptionCode.INVALID_PIN
        QuickCodeException.PinCancelled -> MQuickCodeExceptionCode.PIN_CANCELLED
        QuickCodeException.Revoked -> MQuickCodeExceptionCode.REVOKED
        QuickCodeException.UnsuccessfulAuthentication -> MQuickCodeExceptionCode.UNSUCCESSFUL_AUTHENTICATION
    }

internal val AuthenticationSessionException.flutterExceptionCodeRepresentation: MAuthenticationSessionDetailsExceptionCode
    get() = when (this) {
        is AuthenticationSessionException.AbortSessionFail -> MAuthenticationSessionDetailsExceptionCode.ABORT_SESSION_FAIL
        is AuthenticationSessionException.GetAuthenticationSessionDetailsFail -> MAuthenticationSessionDetailsExceptionCode.GET_AUTHENTICATION_SESSION_DETAILS_FAIL
        AuthenticationSessionException.InvalidAppLink -> MAuthenticationSessionDetailsExceptionCode.INVALID_LINK
        AuthenticationSessionException.InvalidNotificationPayload -> MAuthenticationSessionDetailsExceptionCode.INVALID_NOTIFICATION_PAYLOAD
        AuthenticationSessionException.InvalidQRCode -> MAuthenticationSessionDetailsExceptionCode.INVALID_QRCODE
        AuthenticationSessionException.InvalidSessionDetails -> MAuthenticationSessionDetailsExceptionCode.INVALID_AUTHENTICATION_SESSION_DETAILS
    }

internal val SigningSessionException.flutterExceptionCodeRepresentation: MSigningSessionDetailsExceptionCode
    get() = when (this) {
        is SigningSessionException.AbortSigningSessionFail -> MSigningSessionDetailsExceptionCode.ABORT_SIGNING_SESSION_FAIL
        is SigningSessionException.CompleteSigningSessionFail -> MSigningSessionDetailsExceptionCode.COMPLETE_SIGNING_SESSION_FAIL
        is SigningSessionException.GetSigningSessionDetailsFail -> MSigningSessionDetailsExceptionCode.GET_SIGNING_SESSION_DETAILS_FAIL
        SigningSessionException.InvalidAppLink -> MSigningSessionDetailsExceptionCode.INVALID_LINK
        SigningSessionException.InvalidQRCode -> MSigningSessionDetailsExceptionCode.INVALID_QRCODE
        SigningSessionException.InvalidSigningSession -> MSigningSessionDetailsExceptionCode.INVALID_SIGNING_SESSION
        SigningSessionException.InvalidSigningSessionDetails -> MSigningSessionDetailsExceptionCode.INVALID_SIGNING_SESSION_DETAILS
    }

internal val SigningException.flutterExceptionCodeRepresentation: MSigningExceptionCode
    get() = when (this) {
        SigningException.EmptyMessageHash -> MSigningExceptionCode.EMPTY_MESSAGE_HASH
        SigningException.EmptyPublicKey -> MSigningExceptionCode.EMPTY_PUBLIC_KEY
        SigningException.InvalidPin -> MSigningExceptionCode.INVALID_PIN
        SigningException.InvalidSigningSession -> MSigningExceptionCode.INVALID_SIGNING_SESSION
        SigningException.InvalidSigningSessionDetails -> MSigningExceptionCode.INVALID_SIGNING_SESSION_DETAILS
        SigningException.InvalidUserData -> MSigningExceptionCode.INVALID_USER_DATA
        SigningException.PinCancelled -> MSigningExceptionCode.PIN_CANCELLED
        SigningException.Revoked -> MSigningExceptionCode.REVOKED
        is SigningException.SigningFail -> MSigningExceptionCode.SIGNING_FAIL
        SigningException.UnsuccessfulAuthentication -> MSigningExceptionCode.UNSUCCESSFUL_AUTHENTICATION
    }

File: lib/src/extensions.dart
part of 'miracl_trust.dart';

extension on PlatformException {
  Object? _getExceptionCode() {
    if (details is Map<Object?, Object?>) {
        return details["exceptionCode"];
    }

    return null;
  }
}

extension on MAuthenticationSessionDetails {
  AuthenticationSessionDetails _toAuthenticationSessionDetails() {
    return AuthenticationSessionDetails._create(
        userId: userId, 
        projectName: projectName, 
        projectLogoURL: projectLogoURL, 
        projectId: projectId, 
        pinLength: pinLength,
        verificationMethod: VerificationMethod.getVerificationMethod(verificationMethod.name), 
        verificationURL: verificationURL,
        verificationCustomText: verificationCustomText, 
        identityTypeLabel: identityTypeLabel, 
        quickCodeEnabled: quickCodeEnabled,
        identityType: IdentityType.getIdentityType(identityType.name),
        accessId: accessId
    );
  }
}

extension on MEmailVerificationResponse {
  EmailVerificationResponse _toEmailVerificationResponse(){
    final emailVerificationMethodPublic = EmailVerificationMethod.getEmailVerificationMethod(
      emailVerificationMethod.name
    );

    return  EmailVerificationResponse._create(
      backoff: backoff, 
      emailVerificationMethod: emailVerificationMethodPublic
    );
  }
}

extension on MActivationTokenResponse {
  ActivationTokenResponse _toActivationTokenResponse(){
    return ActivationTokenResponse._create(
        projectId: projectId, 
        accessId: accessId,
        userId: userId, 
        activationToken: activationToken
    );
  }
}

extension on MActivationTokenErrorResponse {
  ActivationTokenErrorResponse _toActivationTokeErrorResponse(){
    return ActivationTokenErrorResponse._create(
      projectId: projectId,
      accessId: accessId,
      userId: userId
    );
  }
}

extension on MUser {
  User _toUser(){
    return User._create(
        userId: userId,
        projectId: projectId,
        revoked: revoked,
        pinLength: pinLength,
        hashedMpinId: hashedMpinId
    );
  }
}

extension on User {
  MUser _toMUser(){
    return MUser(
      projectId: projectId,
      revoked: revoked,
      userId: userId,
      pinLength: pinLength,
      hashedMpinId: hashedMpinId
    );
  }
}

extension on MQuickCode {
  QuickCode _toQuickCode(){
    return QuickCode._create(
      code: code,
      expiryTime: expiryTime,
      ttlSeconds: ttlSeconds
    );
  }
}

extension on MSigningSessionDetails {
  SigningSessionDetails _toSigningSessionDetails() {
    return SigningSessionDetails._create(
        userId: userId,
        projectName: projectName,
        projectLogoURL: projectLogoURL,
        projectId: projectId, 
        pinLength: pinLength, 
        verificationMethod: VerificationMethod.getVerificationMethod(verificationMethod.name),
        verificationURL: verificationURL,
        verificationCustomText: verificationCustomText,
        identityTypeLabel: identityTypeLabel, 
        quickCodeEnabled: quickCodeEnabled,
        identityType: IdentityType.getIdentityType(identityType.name), 
        sessionId: sessionId, 
        signingHash: signingHash, 
        signingDescription: signingDescription, 
        status: SigningSessionStatus.getSigningSessionStatus(status.name),
        expireTime: expireTime
      );
  }
}

extension on MSigningResult {
  SigningResult _toSigningResult() {
    final signatureForReturn = Signature._create(
      u: signature.u, 
      v: signature.v, 
      dtas: signature.dtas, 
      mpinId: signature.mpinId, 
      hash: signature.hash, 
      publicKey: signature.publicKey
    );

    final signingResult = SigningResult._create(
      signature: signatureForReturn,
      timestamp: timestamp
    );

    return signingResult;
  }
}

extension on MConfigurationExceptionCode {
  ConfigurationExceptionCode toConfigurationExceptionCode() => switch (this) {
    MConfigurationExceptionCode.emptyProjectId => ConfigurationExceptionCode.emptyProjectId,
    MConfigurationExceptionCode.invalidProjectUrl => ConfigurationExceptionCode.invalidProjectUrl
  };
}


extension on MEmailVerificationExceptionCode {
  EmailVerificationExceptionCode toEmailVerificationExceptionCode() => switch(this) {
    MEmailVerificationExceptionCode.emptyUserId => EmailVerificationExceptionCode.emptyUserId,
    MEmailVerificationExceptionCode.invalidSessionDetails => EmailVerificationExceptionCode.invalidSessionDetails,
    MEmailVerificationExceptionCode.requestBackoff => EmailVerificationExceptionCode.requestBackoff,
    MEmailVerificationExceptionCode.verificaitonFail => EmailVerificationExceptionCode.verificaitonFail,
  };
}

extension on MActivationTokenExceptionCode {
  ActivationTokenExceptionCode toActivationTokenExceptionCode() => switch(this) {
    MActivationTokenExceptionCode.emptyUserId => ActivationTokenExceptionCode.emptyUserId,
    MActivationTokenExceptionCode.emptyVerificationCode => ActivationTokenExceptionCode.emptyVerificationCode,
    MActivationTokenExceptionCode.unsuccessfulVerification => ActivationTokenExceptionCode.unsuccessfulVerification,
    MActivationTokenExceptionCode.getActivationTokenFail => ActivationTokenExceptionCode.getActivationTokenFail,
  };
}

extension on MRegistrationExceptionCode {
  RegistrationExceptionCode toRegistrationExceptionCode() => switch (this) {
    MRegistrationExceptionCode.emptyUserId => RegistrationExceptionCode.emptyUserId,
    MRegistrationExceptionCode.emptyActivationToken => RegistrationExceptionCode.emptyActivationToken,
    MRegistrationExceptionCode.invalidActivationToken => RegistrationExceptionCode.invalidActivationToken,
    MRegistrationExceptionCode.registrationFail => RegistrationExceptionCode.registrationFail,
    MRegistrationExceptionCode.unsupportedEllipticCurve => RegistrationExceptionCode.unsupportedEllipticCurve,
    MRegistrationExceptionCode.pinCancelled => RegistrationExceptionCode.pinCancelled,
    MRegistrationExceptionCode.invalidPin => RegistrationExceptionCode.invalidPin,
    MRegistrationExceptionCode.projectMismatch => RegistrationExceptionCode.projectMismatch,
  };
}

extension on MAuthenticationExceptionCode {
  AuthenticationExceptionCode toAuthenticationExceptionCode() => switch (this) {
    MAuthenticationExceptionCode.invalidUserData => AuthenticationExceptionCode.invalidUserData,
    MAuthenticationExceptionCode.invalidQRCode => AuthenticationExceptionCode.invalidQRCode,
    MAuthenticationExceptionCode.invalidPushNotificationPayload => AuthenticationExceptionCode.invalidPushNotificationPayload,
    MAuthenticationExceptionCode.userNotFound => AuthenticationExceptionCode.userNotFound,
    MAuthenticationExceptionCode.invalidLink => AuthenticationExceptionCode.invalidLink,
    MAuthenticationExceptionCode.authenticationFail => AuthenticationExceptionCode.authenticationFail,
    MAuthenticationExceptionCode.revoked => AuthenticationExceptionCode.revoked,
    MAuthenticationExceptionCode.invalidAuthenticationSession => AuthenticationExceptionCode.invalidAuthenticationSession,
    MAuthenticationExceptionCode.unsuccessfulAuthentication => AuthenticationExceptionCode.unsuccessfulAuthentication,
    MAuthenticationExceptionCode.pinCancelled => AuthenticationExceptionCode.pinCancelled,
    MAuthenticationExceptionCode.invalidPin => AuthenticationExceptionCode.invalidPin,
    MAuthenticationExceptionCode.invalidCrossDeviceSession => AuthenticationExceptionCode.invalidCrossDeviceSession
  };
}

extension on MQuickCodeExceptionCode {
  QuickCodeExceptionCode toQuickCodeExceptionCode() => switch (this) {
    MQuickCodeExceptionCode.revoked => QuickCodeExceptionCode.revoked,
    MQuickCodeExceptionCode.unsuccessfulAuthentication => QuickCodeExceptionCode.unsuccessfulAuthentication,
    MQuickCodeExceptionCode.pinCancelled => QuickCodeExceptionCode.pinCancelled,
    MQuickCodeExceptionCode.invalidPin => QuickCodeExceptionCode.invalidPin,
    MQuickCodeExceptionCode.generationFail => QuickCodeExceptionCode.generationFail,
  };
}

extension on MAuthenticationSessionDetailsExceptionCode {
  AuthenticationSessionDetailsExceptionCode toAuthenticationSessionDetailsExceptionCode() => switch (this) {
    MAuthenticationSessionDetailsExceptionCode.invalidLink => AuthenticationSessionDetailsExceptionCode.invalidLink,
    MAuthenticationSessionDetailsExceptionCode.invalidQRCode => AuthenticationSessionDetailsExceptionCode.invalidQRCode,
    MAuthenticationSessionDetailsExceptionCode.invalidNotificationPayload => AuthenticationSessionDetailsExceptionCode.invalidNotificationPayload,
    MAuthenticationSessionDetailsExceptionCode.invalidAuthenticationSessionDetails => AuthenticationSessionDetailsExceptionCode.invalidAuthenticationSessionDetails,
    MAuthenticationSessionDetailsExceptionCode.getAuthenticationSessionDetailsFail => AuthenticationSessionDetailsExceptionCode.getAuthenticationSessionDetailsFail,
    MAuthenticationSessionDetailsExceptionCode.abortSessionFail => AuthenticationSessionDetailsExceptionCode.abortSessionFail,
  };
}

extension on MSigningSessionDetailsExceptionCode {
  SigningSessionDetailsExceptionCode toSigningSessionDetailsExceptionCode() => switch (this) {
    MSigningSessionDetailsExceptionCode.invalidLink => SigningSessionDetailsExceptionCode.invalidLink,
    MSigningSessionDetailsExceptionCode.invalidQRCode => SigningSessionDetailsExceptionCode.invalidQRCode,
    MSigningSessionDetailsExceptionCode.invalidSigningSessionDetails => SigningSessionDetailsExceptionCode.invalidSigningSessionDetails,
    MSigningSessionDetailsExceptionCode.getSigningSessionDetailsFail => SigningSessionDetailsExceptionCode.getSigningSessionDetailsFail,
    MSigningSessionDetailsExceptionCode.invalidSigningSession => SigningSessionDetailsExceptionCode.invalidSigningSession,
    MSigningSessionDetailsExceptionCode.completeSigningSessionFail => SigningSessionDetailsExceptionCode.completeSigningSessionFail,
    MSigningSessionDetailsExceptionCode.abortSigningSessionFail => SigningSessionDetailsExceptionCode.abortSigningSessionFail,
  };
}

extension on MSigningExceptionCode {
  SigningExceptionCode toSigningExceptionCode() => switch (this) {
    MSigningExceptionCode.emptyMessageHash => SigningExceptionCode.emptyMessageHash,
    MSigningExceptionCode.emptyPublicKey => SigningExceptionCode.emptyPublicKey,
    MSigningExceptionCode.invalidUserData => SigningExceptionCode.invalidUserData,
    MSigningExceptionCode.pinCancelled => SigningExceptionCode.pinCancelled,
    MSigningExceptionCode.invalidPin => SigningExceptionCode.invalidPin,
    MSigningExceptionCode.signingFail => SigningExceptionCode.signingFail,
    MSigningExceptionCode.revoked => SigningExceptionCode.revoked,
    MSigningExceptionCode.unsuccessfulAuthentication => SigningExceptionCode.unsuccessfulAuthentication,
    MSigningExceptionCode.invalidSigningSession => SigningExceptionCode.invalidSigningSession,
    MSigningExceptionCode.invalidSigningSessionDetails => SigningExceptionCode.invalidSigningSessionDetails,
  };
}
